keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+
+
+Obviously, the default values should validate. But OpenAPI validation schemata do not enforce that. Also one default in a branch of `anyOf`, `allOf`, `oneOf` might force another validation run onto another branch of these quantors.
+
+Even worse: forcing onto another branch might make more defaults applicable. In other words, we would need to run validation with defaulting until it converges. This is a hint that the semantics of defaulting in full OpenAPI schemata are not well defined either.
+
+**We propose to allow defaults only outside of `anyOf`, `allOf`, `oneOf`, `not`.**
+
+Without that, we would get ambiguous defaulting semantics:
+
+#### Example 4:
+
+> Assume 
+>```json
+>{
+>    "anyOf":[
+>       {"default": 1},
+>       {"default": 2}
+>    ]
+>}
+>```
+>
+> Which branch should be chosen while applying the defaults.
+
+#### Example 5:
+
+> Assume
+>
+>```json
+>{
+>  "anyOf":[
+>      {"properties": {"a": {"pattern": "1"}, "b": {"type": "string"}, "d": {"default": 1}}},
+>      {"properties": {"a": {"pattern": "2"}, "c": {"type": "string"}, "d": {"default": 2}}}
+>  ]
+>}
+>```
+>
+> Take an object `{"a": "1", "b":"foo", "c":"bar"}`. Assuming we follow pruning option 1, it is pruned to `{"a": "1", "b":"foo", "c":"bar"}` and to default to `{"a":"1", "b": "foo", "c":"bar", "d": 1}`.
+>
+> But what about the object `{"a": "12", "b":"foo", "c":"bar"}`? Both patterns apply. With pruning option 1 we keep `"b"` and `"c"`. But what about the default? Should we default to `1` or `2`?
+
+Clearly, we can make the algorithm deterministic by choosing one, e.g. the first matching branch. This choice is arbitrary, hard to understand, and gets even more confusing with even more nested `allOf`, `anyOf` or even `not`.
+
+We can later become less restrictive about the allowed positions of default values in the scheme. I.e. the restriction of only allowing default outside of `allOf`, `anyOf`, `oneOf` and `not` is a starting point with a way forward if necessary.
+
+### Mixing of Schema and Value Validation
+
+The OpenAPI validation schema mixes the actual structural schema validation (which is usually done by the Golang struct JSON decoding for native types) 
+and the value validation (usually done in the validation step of the API server handler pipeline for native types). 
+
+For CRDs we cannot distinguish both. This was first noticed by @lavalamp in https://github.com/kubernetes/kubernetes/pull/64907#issuecomment-397015030.
+
+#### Example 6
+
+> Assume 
+>
+>```json
+>{
+>   "properties": {
+>       "a": {"type": "string", "pattern": "<some-long-regex-for-ips>", "format": "ip"},
+>       "b": {"properties": {"x": {}, "y": {}}}
+>   }
+>}
+>```
+>
+> The type and properties are usually called schema validation, while regex patterns and formats are about values. The latter would be tested in the validation phase, the former would be checked during decoding for native types in the JSON decoder.
```

Comment:
**Mehdy Bohlool:** I am not sure about format. it usually uses as an extra identifier with type and may not actually validate the value. Though the border here is blurry. When we say the type is string and the format is date, is it considered a validation or it is the "date" type? My understanding was "type" is primary type and "format" is secondary type.

**Stefan Schimanski:** Added a paragraph below about our discussion around keeping type and format in the skeleton.

Reply:
Right. It's mostly about getting similar error messages.

I would claim we have a general, high-level goal for CRDs to behave like native resources. This is not special for pruning and defaulting. Merely, this goal is one driver to implement pruning and defaulting (and many other features we are adding to CRDs).